From 07dc0a64a3256151beb20a2b4ebc77ee74ebe0dc Mon Sep 17 00:00:00 2001
From: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date: Mon, 28 Sep 2026 01:23:15 +0200
Subject: [PATCH] realloc: Fix mmap non-mremap reallocation case [BZ #34697]

This case was missing copying SIZE_SZ bytes from the old allocation.

Fixes commit 614cfd0f8a2820aed54f9745077c7da0e6643bac
("malloc: Change mmap chunk layout"), which went into glibc 2.43.

Checked by Wilco Dijkstra <Wilco.Dijkstra@arm.com>

Resolves #34697.

(cherry picked from commit 1e0512a684809e32170eae42c8f04f61126d77cc)
---
 malloc/malloc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/malloc/malloc.c b/malloc/malloc.c
index 8fe8b18340..a50ff3598c 100644
--- a/malloc/malloc.c
+++ b/malloc/malloc.c
@@ -3256,7 +3256,7 @@ __libc_realloc (void *oldmem, size_t bytes)
       if (newmem == NULL)
         return NULL;              /* propagate failure */
 
-      memcpy (newmem, oldmem, oldsize - CHUNK_HDR_SZ);
+      memcpy (newmem, oldmem, usable);
       munmap_chunk (oldp);
       return newmem;
     }
-- 
2.53.0

